### PR TITLE
Turn ``SchemaSerializer`` into ``SchemaRegistry``

### PR DESCRIPTION
**Pull request by MattToast**
Rename `SchemaSerializer` to `SchemaRegistry`. Move Message formatting into a dedicated `_Message` class. Rename methods. Better error handling.

### CODE REVIEW

File: smartsim/_core/schemas/utils.py
Comment (al-rigazzi) [inline smartsim/_core/schemas/utils.py]: As I'm not the best candidate to judge English grammar, I deferred to ChatGPT, which says

> The sentence "Failed determine schema type of string" is not grammatically correct. It's missing a verb to connect "Failed" with "determine." To make the sentence grammatically correct, you could use one of the following options:
> - "Failed to determine the schema type of the string."
> - "Failed in determining the schema type of the string."
> 
> Adding "to" before "determine" and including "the" before "schema" and "string" makes the sentence more precise and clear.

**Review by al-rigazzi — CHANGES_REQUESTED** [review]
One grammar nitpick to go!

File: smartsim/_core/schemas/utils.py
Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: 🤦 At this rate, no one is going to believe the English is my native language. Totally correct, will change!

File: smartsim/_core/schemas/utils.py
Comment (ankona) [inline smartsim/_core/schemas/utils.py:54]: What do you think about using `__init__` instead of `from_str`?

I don't think it works after thinking about it... Closing.

File: smartsim/_core/schemas/utils.py
Comment (ankona) [inline smartsim/_core/schemas/utils.py]: nice catch!

File: smartsim/_core/schemas/utils.py
Comment (ankona) [inline smartsim/_core/schemas/utils.py]: I think we might benefit from some `Field` definitions here that enforce non-zero length content for all properties.

It could maybe also remove the questionable validation and "default delimiter" being held in the registry.

Maybe some tests to validate wire format (e.g. if i split, do i get a non-empty lhs/rhs, if i create a new one (y) from a stringified one (str(x)), does str(x) == str(y))...

File: smartsim/_core/entrypoints/dragon.py
Comment (ankona) [inline smartsim/_core/entrypoints/dragon.py]: I wonder if we can just put a thin abstraction over the socket... I think we might avoid forcing callers to do their own conversions?

```py
def send_message(message: Message) -> None:
  wire_format = registry.to_string(Message)
  dragon_head_socket.send_string(wire_format)
```

and callers can feel more natural w/

```py
send_message(resp)
```

i see that the registries are separate, but maybe we should do a little registry unification step:

```py
# silly pseudocode to illustrate the idea of making a `ChainMap` of the existing registry lookup tables
mega_reg.__dict = {**req_reg.__dict, **resp_reg.__dict}
```

File: smartsim/_core/schemas/utils.py
Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: does the header need to be non-empty? I see no problem with doing
```py
@registry.register('')  # <-- Admittedly this will make debugging hell,
                        #     but it shouldn't break anything... I guess that alone is
                        #     enough of a reason to enforce the constraint ¯\_(ツ)_/¯
class MySchema(pydantic.BaseModel):
    ...
```
I'll add the restriction for both! I guess it is easier to add options rather than take them away down the line!!

Comment (ankona) [reply smartsim/_core/schemas/utils.py]: register is taking a "mapping key" which is technically unrelated to the schema, so I feel ok having separate validation. I'm with you...

or... you could perform validation using `_ = Message(key, ...)` and passing it off to the `Message`?

Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: For the "default delimiter", the registry needs to know what it the delimiter is for the messages that it is responsible for formatting/hydrating because it needs to be able to parse out the type information in the header in order to index its `_map`. I can add default to the `_Message` class, but a the Schema registry will still need to always supply its own, and because `_Message` is scoped to this module it will never be used.

I think this conflict is leading to the "questionable validation" you mentioned, but if you could elaborate a little more I should be able remedy/justify!

Comment (ankona) [reply smartsim/_core/schemas/utils.py]: I think this is where that two-way style registry came into the prior convo. We have:

```py
def register(key, type):
  self._map[key] = type  
```

but could add:

```py
def register(key, type):
  self._forward[key] = type
  self._backward[type] = key
```

I think that might let us avoid having to "find" the mapping key in the message within the registry?

Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: How is `_backward` different than `reverse_map` in `_to_message`? All that would do is save some iteration cycles when calling, `_to_message`, but given that we are dealing with relatively small dicts, I'm not sure I want to add the complexity overhead of making sure `_forward` and `_backward` stay "in sync".

To clarify my previous comment, the delimiter is not needed to index into `_map` (or indeed `_forward`). It is needed to parse out the type information of the `str_` parameter of `from_string` in order to know what to index _with_.

Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: For reference, [this is the offending line](https://github.com/al-rigazzi/SmartSim/pull/8/files#diff-eb87d606b148ad7705ae9e182b6f5e555ace51718f6a2dce191994ca9f8213e2R100)

**Review by al-rigazzi — APPROVED** [review]
Like it! LGTM, pending @ankona's approval.

File: smartsim/_core/entrypoints/dragon.py
Comment (MattToast) [reply smartsim/_core/entrypoints/dragon.py]: Makes sense to me!! I didn't love the idea of having a "mega-registry" for two main reasons: (a) it would remove a lot of the type narrowing benefits of having specific registries for specific purposes, and (b) I couldn't come up with a proper way to handle key conflicts on both ends of the wire seamlessly. That said, I see no reason our abstraction cannot simply use two registries: one for incoming and one for out going messages!

Added the `SocketSchemaTranslator` class to handle conversions of schemas to and from strings for sockets, anded some helper fns for dragon specific use of the `SocketSchemaTranslator` class, removed the `_Pipline` usage as it became obsolete, and added tests!

**Review by ankona — APPROVED** [review]
LGTM!

File: smartsim/_core/schemas/utils.py
Comment (MattToast) [reply smartsim/_core/schemas/utils.py]: Per offline conversation, moved `_DEFAULT_MSG_DELIM` to a private module global. This removes the need to make it an attr on the `SchemaRegistry` and because it is private, we can always choose to expose/change it later!

**Pull request merged**
